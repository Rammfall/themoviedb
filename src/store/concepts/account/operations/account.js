import { createLogic } from 'redux-logic'

import storage from 'Utils/storage'
import { setUserId, setUsername } from 'Store/concepts/account/actions'
import { GET_INFO } from '../types'
import { account } from '../endpoints'

const getAccountOperation = createLogic({
  type: GET_INFO,
  latest: true,
  async process(
    {
      httpClient
    },
    dispatch,
    done
  ) {
    const { data: { username, id } } = await httpClient.get(account, { params: { session_id: storage.session.get() } })

    dispatch(setUsername({ username }))
    dispatch(setUserId({ id }))

    done()
  }
})

export default getAccountOperation
