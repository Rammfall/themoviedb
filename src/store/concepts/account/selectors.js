export const usernameSelector = ({ account: { username } }) => username

export const userIdSelector = ({ account: { userId } }) => userId
