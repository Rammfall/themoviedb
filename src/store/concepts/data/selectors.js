import { createSelector } from 'reselect'

const dataSelector = state => state.data

export const loadingSelector = createSelector(
  dataSelector,
  (_, endpoint) => endpoint,
  (data, endpoint) => data[endpoint] && data[endpoint].isLoading
)
