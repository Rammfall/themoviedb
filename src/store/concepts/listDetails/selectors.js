import { createSelector } from 'reselect'

/**
 * @returns {{total: number, isEmpty: boolean, movies: Object[]}}
 */
export const getListDetailsMoviesSelector = createSelector(
  ({ data: { movies } }) => movies,
  ({ data: { lists } }) => lists,
  ({ listDetails }) => listDetails,
  (_, id) => id,
  (movies, lists, listsDetails, id) => ({
    movies: listsDetails[id].ids.map((item) => movies[item]),
    isEmpty: !listsDetails[id].ids.length,
    total: listsDetails[id].total
  })
)

/**
 * @returns {string}
 */
export const getListTitleSelector = createSelector(
  ({ data: { lists } }) => lists,
  (_, id) => id,
  (lists, id) => lists[id].name
)

/**
 * @returns {boolean}
 */
export const isBlankSelector = createSelector(
  ({ listDetails }) => listDetails,
  (_, id) => id,
  (listDetails, id) => !listDetails[id]
)
