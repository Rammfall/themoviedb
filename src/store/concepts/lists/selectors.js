import { createSelector } from 'reselect'

const dataSelector = (state) => state.data.lists

const listsIdsSelector = ({ lists: { listsIds } }) => listsIds

export const getListsSelector = createSelector(
  dataSelector,
  listsIdsSelector,
  (lists, listsIds) => listsIds.map(item => lists[item])
)

export const getListsTotalSelector = ({ lists: { listsTotal } }) => listsTotal

export const isEmptyListsSelector = createSelector(
  listsIdsSelector,
  (ids) => !ids.length
)
