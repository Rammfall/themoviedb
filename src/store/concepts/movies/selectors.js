import { createSelector } from 'reselect'

import { moviesConstant } from 'Constants/concepts'

const dataSelector = (state) => state.data.movies

const dashboardIdsSelector = ({ movies: { dashboardIds } }) => dashboardIds

export const getDashboardMoviesSelector = createSelector(
  dataSelector,
  dashboardIdsSelector,
  (movies, moviesIds) => moviesIds.map(item => movies[item])
)

export const getDashboardMoviesTotalSelector = ({ movies: { dashboardTotal } }) => dashboardTotal

export const isEmptyDashboardSelector = createSelector(
  dashboardIdsSelector,
  (ids) => !ids.length
)

const watchlistIdsSelector = ({ movies: { watchlist: { ids } } }) => ids
const watchlistTotalSelector = ({ movies: { watchlist: { total } } }) => total

export const getWatchlistMoviesSelector = createSelector(
  dataSelector,
  watchlistIdsSelector,
  watchlistTotalSelector,
  (movies, ids, total) => ({ total, isEmpty: !ids.length, movies: ids.map((item) => movies[item]) })
)
// Try to use createStructuredSelector and replace map to another selector

const favoritesIdsSelector = ({ movies: { favorites: { ids } } }) => ids
const favoritesTotalSelector = ({ movies: { favorites: { total } } }) => total

export const getFavoritesMoviesSelector = createSelector(
  dataSelector,
  favoritesIdsSelector,
  favoritesTotalSelector,
  (movies, ids, total) => ({ total, isEmpty: !ids.length, movies: ids.map((item) => movies[item]) })
)

const listsSelector = ({ data: { lists } }) => lists

export const getListMoviesSelector = createSelector(
  dataSelector,
  listsSelector,
  (_, id) => id,
  (movies, lists, id) => {
    const listMovies = lists[id][moviesConstant]

    return {
      total: listMovies.total,
      isEmpty: !listMovies.ids.length,
      movies: listMovies.ids.map((item) => movies[item])
    }
  }
)
