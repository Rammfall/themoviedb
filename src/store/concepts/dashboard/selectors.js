import { createSelector } from 'reselect'

import { idsSelector, moviesSelector, shapeSelector, totalSelector } from 'Store/selectors'
import { DASHBOARD, MOVIES } from 'Constants/concepts'

/**
 * @returns {{movies: Object[], total: number, isEmpty: boolean}}
 */
export const getDashboardMoviesSelector = createSelector(
  moviesSelector,
  idsSelector(DASHBOARD),
  totalSelector(DASHBOARD),
  shapeSelector(MOVIES)
)
