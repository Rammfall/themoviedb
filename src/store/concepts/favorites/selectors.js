import { createSelector } from 'reselect'

import { idsSelector, moviesSelector, shapeSelector, totalSelector } from 'Store/selectors'
import { FAVORITES, MOVIES } from 'Constants/concepts'

export const getFavoritesMoviesSelector = createSelector(
  moviesSelector,
  idsSelector(FAVORITES),
  totalSelector(FAVORITES),
  shapeSelector(MOVIES)
)
