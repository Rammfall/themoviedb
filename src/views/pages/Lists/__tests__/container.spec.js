import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import diveTo from 'TestUtils/diveToEnzyme'

import { userIdSelector } from 'Store/concepts/account/selectors'

import ListsPage, { ListsPageContainer } from '../container'

jest.mock('Store/concepts/lists/selectors', () => ({
  isEmptyListsSelector: jest.fn(() => false)
}))

jest.mock('Store/concepts/account/selectors', () => ({
  userIdSelector: jest.fn(() => null)
}))

jest.mock('Store/concepts/data/selectors', () => ({
  loadingSelector: jest.fn()
}))

jest.mock('react-router-dom', () => ({
  withRouter: component => component
}))

describe('ListsPage', () => {
  const store = configureStore()()
  const props = {
    location: {
      search: ''
    }
  }
  const wrapper = shallow(
    <ListsPage
      store={store}
      {...props}
    />
  )

  describe('with default props', () => {
    it('matches snapshot', () => {
      const container = diveTo(wrapper, ListsPageContainer)

      expect(container).toMatchSnapshot()
    })

    describe('componentDidUpdate', () => {
      it('have been called', () => {
        const container = diveTo(wrapper, ListsPageContainer)
        const spy = jest.spyOn(container.instance(), 'componentDidUpdate')
        container.setProps({ location: { search: '?page=3' } })

        expect(spy).toHaveBeenCalled()
      })
    })

    describe('toggleCreateModal', () => {
      it('matches snapshot', () => {
        const container = diveTo(wrapper, ListsPageContainer)
        const spy = jest.spyOn(container.instance(), 'toggleCreateModal')
        container.instance().toggleCreateModal()

        expect(spy).toHaveBeenCalled()
      })
    })
  })

  describe('when userId is null', () => {
    userIdSelector.mockImplementation(() => null)
    it('matches snapshot', () => {
      const container = diveTo(wrapper, ListsPageContainer)

      expect(container).toMatchSnapshot()
    })

    describe('after componentDidUpdate', () => {
      describe('when keyword props was not changed', () => {
        it('matches snapshot', () => {
          const container = diveTo(wrapper, ListsPageContainer)
          container.setProps({ test: 'tes' })

          expect(container).toMatchSnapshot()
        })
      })

      describe('when search was changed', () => {
        it('matches snapshot', () => {
          const container = diveTo(wrapper, ListsPageContainer)
          container.setProps({ location: { search: '?page=53' } })

          expect(container).toMatchSnapshot()
        })
      })

      describe('when userId was changed', () => {
        it('matches snapshot', () => {
          const container = diveTo(wrapper, ListsPageContainer)
          container.setProps({ userId: 4 })

          expect(container).toMatchSnapshot()
        })
      })
    })
  })
})
