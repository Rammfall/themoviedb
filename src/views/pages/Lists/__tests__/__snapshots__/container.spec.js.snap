// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsPage when userId is null after componentDidUpdate when keyword props was not changed matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage when userId is null after componentDidUpdate when search was changed matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage when userId is null after componentDidUpdate when userId was changed matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage when userId is null matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage with default props after componentDidUpdate matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage with default props after setState matches snapshot 1`] = `
<Lists
  createModalVisibility={true}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;

exports[`ListsPage with default props matches snapshot 1`] = `
<Lists
  createModalVisibility={false}
  isEmpty={false}
  isLoading={true}
  toggleCreateModal={[Function]}
/>
`;
